provider/equinix/environ.go

```diff
@@ -0,0 +1,693 @@
+			requestedPrivateAddr = requestedPrivateAddr || !net.Public
+
+			// Packet requires us to request at least a /31 for IPV4
+			// addresses and a /127 for IPV6 ones.
+			cidrSize := 31
+			if net.AddressFamily != 4 {
+				cidrSize = 127
+			}
+
+			ipBlock := packngo.IPAddressCreateRequest{
+				AddressFamily: net.AddressFamily,
+				Public:        net.Public,
+				CIDR:          cidrSize,
+				Reservations:  []string{net.ID},
+			}
+			device.IPAddresses = append(device.IPAddresses, ipBlock)
+		}
+	}
+
+	// In order to spin up a new device, we must specify at least one
+	// public and one private address.
+	if !requestedPrivateAddr {
+		// Allocate a private address from the default address pool.
+		device.IPAddresses = append(device.IPAddresses, packngo.IPAddressCreateRequest{
+			Public:        false,
+			AddressFamily: 4,
+			CIDR:          31,
+		})
+	}
+	if !requestedPublicAddr {
+		// Allocate a public address from the default address pool.
+		device.IPAddresses = append(device.IPAddresses, packngo.IPAddressCreateRequest{
+			Public:        true,
+			AddressFamily: 4,
+			CIDR:          31,
+		})
+	}
+
+	d, _, err := e.equnixClient.Devices.Create(device)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	d, err = waitDeviceActive(e.equnixClient, d.ID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	inst := &equnixDevice{e, d}
+	amd64 := arch.AMD64
+	mem, err := strconv.ParseUint(d.Plan.Specs.Memory.Total[:len(d.Plan.Specs.Memory.Total)-2], 10, 64)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var cpus uint64 = 1
+	if inst.Plan != nil && inst.Plan.Specs != nil && len(inst.Plan.Specs.Cpus) > 0 {
+		cpus = uint64(inst.Plan.Specs.Cpus[0].Count)
+	}
+
+	return &environs.StartInstanceResult{
+		Instance: inst,
+		Hardware: &instance.HardwareCharacteristics{
+			Arch: &amd64,
+			Mem:  &mem,
+			// RootDisk: &instanceSpec.InstanceType.RootDisk,
+			CpuCores: &cpus,
+		},
+	}, nil
+}
+
+func (e *environ) getSubnetsToZoneMap(ctx context.ProviderCallContext, args environs.StartInstanceParams) ([]string, error) {
+	var subnetIDs []string
+	for _, subnetList := range args.SubnetsToZones {
+		for subnetID := range subnetList {
+			packetSubnetID := strings.TrimPrefix(subnetID.String(), "subnet-")
+			subnetIDs = append(subnetIDs, packetSubnetID)
+		}
+	}
+
+	return subnetIDs, nil
+}
+
+// supportedInstanceTypes returns the instance types supported by Equnix Metal.
+func (e *environ) supportedInstanceTypes() ([]instances.InstanceType, error) {
+	opt := &packngo.ListOptions{
+		Includes: []string{"available_in"},
+	}
+	plans, _, err := e.equnixClient.Plans.List(opt)
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving supported instance types")
+	}
+
+	var instTypes []instances.InstanceType
+nextPlan:
+	for _, plan := range plans {
+		if !validPlan(plan, e.cloud.Region) {
+			logger.Infof("Plan %s not valid in facility %s", plan.Name, e.cloud.Region)
```

Comment:
This should be logged at DEBUG level. Also, from a quick run against EWR1, I see that multiple plans get rejected as non-valid. Any ideas why?

```
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan Type 2A4 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan Type HUA not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan c1.bloomberg.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan c3.medium.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan c3.small.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan cpe1.g1.4028gr not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan cpe1.m2.r640 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan cpe1.s1.r730 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan cpe1.x1.supermicro not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan f3.large.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan f3.medium.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan g2.large.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan m2.xlarge.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan m3.large.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan s3.xlarge.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan Standard not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan Performance not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan Custom Storage Appliance not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan t3.small.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan x3.xlarge.x86 not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan x.large.arm not valid in facility ewr1
13:12:45 INFO  juju.provider.equnix environ.go:478 Plan x.small.x86 not valid in facility ewr1
```

Looks like only the `t1.small.x86` is available. Is that expected?

Reply:
`ewr1` has become an expansion site, reservations only.

`dc13` may be a better choice for On-Deman provisioning.

We may benefit from adapting this from a "facility" based deployment to a "metro" based deployment, to take advantage of broader availability. https://metal.equinix.com/developers/docs/locations/metros/#metros-quick-reference